[SPIR-V] Add SPIR-V structurizer

Keenuts opened the pull request:
This commit  adds an initial SPIR-V structurizer.
It leverages the previously merged passes, and the convergence region analysis to determine the correct merge and continue blocks for SPIR-V.

The first part replaces switches with a single edge with a simple branch.
Then, we add OpLoopMerge instruction to each loop encountered. Then, we add OpSelectionMerge for each conditional branch which has no OpLoopMerge instruction.

Finally, we fixup the merge instructions:
 - we split blocks with multiple headers into 2 blocks.
 - we split blocks that are a merge blocks for 2 or more constructs: SPIR-V spec disallow a merge block to be shared by 2 loop/switch/condition construct.
 - we split merge & continue blocks: SPIR-V spec disallow a basic block to be both a continue block, and a merge block.
 - we remove superfluous headers: when a header doesn't bring more info than the parent on the divergence state, it must be removed.
 - We sort blocks depending on the dominator tree order: SPIR-V spec requires blocks to be sorted in a specific way.

As is, this code seems to pass most DXC structurization tests, except the ones relying on unrelated features this backend doesn't yet support like i64 switches, boolean types.
One known case fails, but it is because the MergeExit pass doesn't supports switches yet (there is a FIXME).
This PR is already big-enough as-is, so I think we should get this in first, and then add support for those switches.

Keenuts committed:
[SPIR-V] Add SPIR-V structurizer

This commit  adds an initial SPIR-V structurizer.
It leverages the previously merged passes, and the convergence region
analysis to determine the correct merge and continue blocks for SPIR-V.

The first part replaces switches with a single edge with a simple
branch.
Then, we add OpLoopMerge instruction to each loop encountered.
Then, we add OpSelectionMerge for each conditional branch which has
no OpLoopMerge instruction.

Finally, we fixup the merge instructions:
 - we split blocks with multiple headers into 2 blocks.
 - we split blocks that are a merge blocks for 2 or more constructs:
        SPIR-V spec disallow a merge block to be shared by 2
        loop/switch/condition construct.
 - we split merge & continue blocks:
        SPIR-V spec disallow a basic block to be both a continue
        block, and a merge block.
 - we remove superfluous headers: when a header doesn't bring more info
   than the parent on the divergence state, it must be removed.
 - We sort blocks depending on the dominator tree order:
        SPIR-V spec requires blocks to be sorted in a specific way.

As is, this code seems to pass most DXC structurization tests, except
the ones relying on unrelated features this backend doesn't yet support
like i64 switches, boolean types.
One known case fails, but it is because the MergeExit pass doesn't
supports switches yet (there is a FIXME).
This PR is already big-enough as-is, so I think we should get this in
first, and then add support for those switches.

michalpaszkowski commented on llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:
```diff
@@ -0,0 +1,649 @@
+// Gather all the successors of |BB|.
+// This function asserts if the terminator neither a branch, switch or return.
+std::unordered_set<BasicBlock *> gatherSuccessors(BasicBlock *BB) {
+  std::unordered_set<BasicBlock *> output;
```
Please start [variable names with upper case letter](https://llvm.org/docs/CodingStandards.html#id44).

michalpaszkowski commented on llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:
```diff
@@ -0,0 +1,649 @@
+// Returns all basic blocks in |MF| with at least one SelectionMerge/LoopMerge
+// instruction.
+SmallPtrSet<MachineBasicBlock *, 8> getHeaderBlocks(MachineFunction &MF) {
+  SmallPtrSet<MachineBasicBlock *, 8> output;
```
Please start [variable names with upper case letter](https://llvm.org/docs/CodingStandards.html#id44).

michalpaszkowski commented on llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:
```diff
@@ -0,0 +1,649 @@
+// Returns all basic blocks in |MF| referenced by at least 1
+// OpSelectionMerge/OpLoopMerge instruction.
+SmallPtrSet<MachineBasicBlock *, 8> getMergeBlocks(MachineFunction &MF) {
+  SmallPtrSet<MachineBasicBlock *, 8> output;
```
Please start [variable names with upper case letter](https://llvm.org/docs/CodingStandards.html#id44).

michalpaszkowski commented on llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:
```diff
@@ -0,0 +1,649 @@
+// Returns all basic blocks in |MF| referenced as continue target by at least 1
+// OpLoopMerge.
+SmallPtrSet<MachineBasicBlock *, 8> getContinueBlocks(MachineFunction &MF) {
+  SmallPtrSet<MachineBasicBlock *, 8> output;
```
Please start [variable names with upper case letter](https://llvm.org/docs/CodingStandards.html#id44).

michalpaszkowski commented on llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:
```diff
@@ -0,0 +1,649 @@
+// Returns the block immediatly post-dominating every block in |range| if any,
+// nullptr otherwise.
+MachineBasicBlock *findNearestCommonDominator(
+    const iterator_range<std::vector<MachineBasicBlock *>::iterator> &range,
```
Please start [variable names with upper case letter](https://llvm.org/docs/CodingStandards.html#id44) (`range`). I think this also applies to arguments, but there are various places in the repository which do not adhere to this.

michalpaszkowski commented on llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:
```diff
@@ -0,0 +1,649 @@
+
+  // Replace switches with a single target with an unconditional branch.
+  bool replaceEmptySwitchWithBranch(MachineFunction &MF) {
+    bool modified = false;
```
Please start [variable names with upper case letter](https://llvm.org/docs/CodingStandards.html#id44).

michalpaszkowski commented on llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:
```diff
@@ -0,0 +1,649 @@
+            .getRegionInfo()
+            .getTopLevelRegion();
+
+    bool modified = false;
```
Please start [variable names with upper case letter](https://llvm.org/docs/CodingStandards.html#id44).

michalpaszkowski commented on llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:
```diff
@@ -0,0 +1,649 @@
+
+    // Maps each merge-block to its associated header block.
+    std::unordered_map<MachineBasicBlock *, MachineBasicBlock *> MergeToHeader;
+    bool modified = false;
```
Please start [variable names with upper case letter](https://llvm.org/docs/CodingStandards.html#id44).

michalpaszkowski commented on llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:
```diff
@@ -0,0 +1,649 @@
+          [&toProcess](MachineBasicBlock *MBB) { toProcess.push_back(MBB); });
+
+    auto ContinueBlocks = getContinueBlocks(MF);
+    bool modified = false;
```
Please start [variable names with upper case letter](https://llvm.org/docs/CodingStandards.html#id44).

michalpaszkowski commented on llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:
```diff
@@ -0,0 +1,649 @@
+      if (MI->getOpcode() == SPIRV::OpLoopMerge)
+        continue;
+
+      size_t dominated_count = 0;
```
Please start [variable names with upper case letter and follow with camel case](https://llvm.org/docs/CodingStandards.html#id44).

michalpaszkowski commented on llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:
```diff
@@ -0,0 +1,649 @@
+  }
+
+  virtual bool runOnMachineFunction(MachineFunction &MF) override {
+    bool modified = false;
```
Please start [variable names with upper case letter](https://llvm.org/docs/CodingStandards.html#id44).

michalpaszkowski commented on llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:
```diff
@@ -0,0 +1,649 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include <stack>
```
Please list system includes at the end. The [expected order can be found here](https://llvm.org/docs/CodingStandards.html#id16).

Keenuts replied:
Arf the opposite than us, fixed thanks!

michalpaszkowski reviewed (commented):
Thank you @Keenuts for the patch! I added some small nitpicky comments, but I think the patch looks great and is really easy to follow!

sudonatalie commented:
@s-perron, can you do a review of the implementation logic here? I can't remember if you're an llvm committer, but if not I'll do a secondary review and stamp if it looks good to you.

Keenuts committed:
style fixes

Keenuts committed:
rebase fixes

Keenuts commented:
> Thank you @Keenuts for the patch! I added some small nitpicky comments, but I think the patch looks great and is really easy to follow!

Thanks for the review! Fixed the style issues and rebased on main.

s-perron commented on llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:
```diff
@@ -0,0 +1,643 @@
+    return Modified;
+  }
+
+  // Add an OpSelectionMerge to each node with an out-degree of 2 or more.
+  bool addMergeForConditionalBranches(MachineFunction &MF) {
```
How do you make sure that all blocks with in-degree of 2 or more are the merge block of a header? The example you give below is a good example?

```
//          A
//         / \
//        B<--C
//        \    \
//         \    D
//          \  /
//           E
```

From what I can tell, this will add two `OpSelectionMerge` instructions. One in `A` and one in `C`. Both will use `E` as the merge block. `B` will not be a merge block. This violates the rules for maximal reconvergence (https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_maximal_reconvergence.html).

Note, we want to use maximal reconvergence now that it is released. Otherwise, many cases would have implementation defined behaviour.

s-perron commented on llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:
```diff
@@ -0,0 +1,643 @@
+
+// Returns the block immediatly post-dominating every block in |Range| if any,
+// nullptr otherwise.
+MachineBasicBlock *findNearestCommonDominator(
```
The comment says you are looking for a node that post dominates the range, but the function names just looks for a dominator. This is confusing because they are different things.

```suggestion
MachineBasicBlock *findNearestCommonPostDominator(
```

s-perron reviewed (changes requested):
Overall looks good, and I like how it is divided into small functions. That is made it easy to review.

However, I have a problem with the step that adds the selection merge instructions. We have disagreed on this general, but I think we need to implement it by looking at the merge blocks. We need to add a selection merge for every block that has multiple predecessors. That selection merge should be the immediate dominator of the block.

https://en.wikipedia.org/wiki/Dominator_(graph_theory)

Keenuts commented on llvm/lib/Target/SPIRV/Analysis/SPIRVConvergenceRegionAnalysis.cpp:
```diff
@@ -230,7 +230,8 @@ class ConvergenceRegionAnalyzer {
       auto *To = Terminator->getSuccessor(i);
-      if (isBackEdge(From, To))
+      // Ignore back edges and self edges.
+      if (From == To || isBackEdge(From, To))
```
Yes, now that you say it... 🙃 Removed the bogus assert in isBackEdge and reverted this.

Keenuts committed:
pr feedback: small nits/typos

Signed-off-by: Nathan Gauër <brioche@google.com>

Keenuts commented:
Closing this PR for now.
Rewrite the structurizer to work from the merge nodes, and also handle some cases with switches and fallthrough. Some issues I encountered outweighed the small benefit I had to make this as the MIR level instead of the IR level. The next PR will be a complete rewrite at the IR level.
Closing this one as a result.
